samples/qtask_ordered.c

```diff
@@ -0,0 +1,349 @@
+    if (!(pool = hts_tpool_init(cnt))) {                //thread pool
+        printf("Failed to create thread pool\n");
+        goto end;
+    }
+    //queue to use with thread pool, for task and results
+    if (!(queue = hts_tpool_process_init(pool, cnt * 2, 0))) {
+        printf("Failed to create queue\n");
+        goto end;
+    }
+    //open input file - r reading
+    if (!(infile = sam_open(inname, "r"))) {
+        printf("Could not open %s\n", inname);
+        goto end;
+    }
+    //open output files - w write as SAM, wb  write as BAM
+    if (!(outfile = sam_open(file, "w"))) {
+        printf("Could not open output file\n");
+        goto end;
+    }
+    //read header, required to resolve the target names to proper ids
+    if (!(in_samhdr = sam_hdr_read(infile))) {
+        printf("Failed to read header from file!\n");
+        goto end;
+    }
+    /*tasks are queued, worker threads get them and processes in parallel;
+    the results are queued and they are to be removed in parallel as well*/
+    //start output writer thread for ordered processing
+    twritedata.outfile = outfile;
+    twritedata.queue = queue;
+    twritedata.done = &stopcond;
+    twritedata.lock = &stopcondsynch;
+    twritedata.samhdr = in_samhdr;
+    if (pthread_create(&thread, NULL, threadfn_orderedwrite, &twritedata)) {
+        printf("Failed to create writer thread\n");
+        goto end;
+    }
+    clearthread = 1;
+    //write header
+    if ((sam_hdr_write(outfile, in_samhdr) == -1)) {
+        printf("Failed to write header\n");
+        goto end;
+    }
+
+    c = 0;
+    while (c >= 0) {
+        bamdata = getbamstorage(chunk);
+        //read alignments, upto max size for this lot
+        for (cnt = 0; cnt < bamdata->size; ++cnt) {
+            c = sam_read1(infile, in_samhdr, bamdata->bamarray[cnt]);
+            if (c >= 0) {
+                continue;   //read next
+            }
+            else {
+                break;      //failure
+            }
+        }
+        if (c >= -1 ) {
+            //max size data or reached EOF
+            bamdata->count = ( c >= 0 )? bamdata->size : cnt;
+            //queue the lot for processing
+            if (hts_tpool_dispatch(pool, queue, thread_ordered_proc,
+                bamdata) == -1) {
+                printf("Failed to schedule processing\n");
+                goto end;
+            }
+            bamdata = NULL;
+        }
+        else {
+            printf("Error in reading data\n");
+            break;
+        }
+    }
+    if (-1 == c) {
+        //EOF read, trigger processing for anything pending, NOTE: will be blocked until queue is cleared
+        if (hts_tpool_process_flush(queue) == -1) {
+            printf("Failed to flush queues\n");
+            goto end;
+        }
+        ret = EXIT_SUCCESS;
+    }
+
+    usleep(WAIT * 1000000);                     //to ensure result check is done atleast once after processing
```

Comment:
Similarly this rings huge alarm bells about race conditions.    I changed it to `*1000` and sure enough on the 5th test is wedged.

To detect ending we need signalling, not waiting and hoping, as I/O could be slow.

Reply:
The worker threads won't be able to signal end, and hence the user thread detects it using empty result queue and explicit exit trigger from main thread. The wait is to ensure that the result check is made at least once before exit.

Reply:
this is replaced with a loop, to check output queue status using hts_tpool_process_empty with 1 sec wait.
now main thread queues tasks, on end, execute flush to finish processing and waits until output queue is empty and then trigger thread exit.